File: Python/ByTier/F/1991_F_Triangle_Formation.py
# Problem: CF 1991 F - Triangle Formation
# https://codeforces.com/contest/1991/problem/F

"""
Problem: Triangle Formation
Purpose: Determine if 2 non-degenerate triangles can be formed from a subarray of sticks.
Algorithms/Techniques: 
- Sorting the subarray to simplify triangle inequality checks
- Greedy approach for forming first triangle
- Brute-force check for second triangle using combinations of 6 elements
Time Complexity: O(n log n + q * n) where n is number of sticks and q is number of queries
Space Complexity: O(n) for storing the array and sorting temporary subarrays
"""

from sys import *

input = stdin.readline
inputlist = lambda: map(int, stdin.readline().split())


def solve():
    l, r = inputlist()
    if r - l + 1 >= 48:
        return "YES"
    else:
        # Sort the subarray to simplify triangle checks
        array = sorted(a[l - 1 : r])
        count = 0
        i = 0
        # Try to form first triangle using greedy approach
        # If three consecutive elements form a valid triangle, we consider it as such
        while i < len(array) - 2:
            if array[i] + array[i + 1] > array[i + 2]:
                count += 1
                i += 3  # Move to next potential triangle after current one
            else:
                i += 1
            if count >= 2:
                return "YES"
        # Check all combinations of choosing 6 sticks to form 2 triangles
        # Enumerate all possible pairs of triangles using 6 indices
        for i in range(len(array) - 5):
            # Triangle 1: indices i, i+1, i+2; Triangle 2: indices i+3, i+4, i+5
            if (
                array[i] + array[i + 2] > array[i + 3]
                and array[i + 1] + array[i + 4] > array[i + 5]
            ):
                return "YES"
            # Triangle 1: indices i, i+1, i+3; Triangle 2: indices i+2, i+4, i+5
            if (
                array[i] + array[i + 3] > array[i + 4]
                and array[i + 1] + array[i + 2] > array[i + 5]
            ):
                return "YES"
            # Triangle 1: indices i, i+1, i+4; Triangle 2: indices i+2, i+3, i+5
            if (
                array[i] + array[i + 4] > array[i + 5]
                and array[i + 1] + array[i + 2] > array[i + 3]
            ):
                return "YES"
    return "NO"


n, q = inputlist()
a = list(inputlist())
for _ in range(q):
    print(solve())


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1991/1991_F_Triangle_Formation.cpp
// Problem: CF 1991 F - Triangle Formation
// https://codeforces.com/contest/1991/problem/F

/*
Purpose: This program determines whether it's possible to choose 6 distinct sticks from a given range [l, r] 
         to form 2 non-degenerate triangles. A non-degenerate triangle satisfies the triangle inequality:
         the sum of any two sides must be greater than the third side.

Algorithms/Techniques:
- Sorting: Sorts the subarray to efficiently test combinations.
- Brute-force checking for triangle combinations: Tries different groupings of sticks to form 2 triangles.
- Optimization for large ranges: For ranges larger than 48 elements, it's always possible to form 2 triangles (output "YES").

Time Complexity: O(n * log(n) + q * n) where n is the number of sticks and q is number of queries.
Space Complexity: O(n) for the array of sticks and auxiliary array for sorting.

Note: The brute-force approach is optimized by limiting the number of combinations checked and using the fact
      that for large enough ranges (more than 48 elements), it's always possible to form 2 triangles.
*/

#include <stdio.h>
#include <algorithm>
#include <iostream>

using namespace std;
int a[100005], c[50], tot;
inline bool check(int a, int b, int c) { return a + b > c; }
inline bool check(int a, int b, int c, int d, int e, int f) {
  return check(a, b, c) && check(d, e, f) || check(a, c, d) && check(b, e, f) ||
         check(a, d, e) && check(b, c, f) || check(a, d, f) && check(b, c, e) ||
         check(a, e, f) && check(b, c, d);
}
int main() {
  int n, q;
  cin >> n >> q;
  for (int i = 1; i <= n; i++) {
    scanf("%d", &a[i]);
  }
  while (q--) {
    int l, r;
    scanf("%d%d", &l, &r);
    if (r - l + 1 < 48) {
      tot = 0;
      for (int i = l; i <= r; i++) {
        c[++tot] = a[i];
      }
      sort(c + 1, c + tot + 1);
      bool ok = false;
      for (int i = 1; i + 5 <= tot; i++) {
        ok |= check(c[i], c[i + 1], c[i + 2], c[i + 3], c[i + 4], c[i + 5]);
      }
      int l = -1, r;
      for (int i = 1; i + 2 <= tot; i++) {
        if (check(c[i], c[i + 1], c[i + 2])) {
          l = i;
          break;
        }
      }
      for (int i = tot - 2; i >= 1; i--) {
        if (check(c[i], c[i + 1], c[i + 2])) {
          r = i;
          break;
        }
      }
      ok |= (l != -1 && l + 2 < r);
      printf("%s\n", (ok ? "YES" : "NO"));
    } else {
      printf("YES\n");
    }
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/